docs/layers/base-layer.md

```diff
@@ -201,45 +183,18 @@ can be refactored to avoid the need.
 
 ##### `numInstances` (Number, optional)
 
-deck.gl is often able to autodetect the number of objects in your `data` prop,
-however in special situations it can be useful to control this directly.
+deck.gl automatically derives the number of drawing instances from the `data` prop by
+counting the number of objects in `data`. However, the developer might want to
+maunally override it using this prop.
 
 ##### `updateTriggers`
 
-This prop expects an object with keys matching attribute names.
-Each attributeName will contain an object with keys representing property names.
-If any property mentioned in the updateProps has changed, the corresponding
-attribute will be invalidated.
-
-This way, if an attribute update depends on properties other than data,
-the layer can ensure that the attribute gets autoupdated when those props
-change.
-
-The value of `updateTriggers` is a map with fields corresponding to
-attribute names (or `all`). Each field has a value which is an object,
-it can contain any amount of data. The data for each field is compared
-shallowly, and if a change is detected, the attribute is invalidated
-(all attributes are invalidated if the `all` key is used.)
-
-Note: updateTriggers are ignored by the normal shallow comparison, so it is
-ok for the app to mint a new object on every render.
-
-##### "attributes" properties
-
-The layer will look for properties with names matching its attributes,
-and if present, assume their contents is a typed array or WebGLBuffer,
-and pass them directly to the shader. This also disables the automatic
-attribute calculation.
-
-This is intended to support applications that want to optimize attribute
-calculation. Perhaps only a few data items change each frame, and recalculating
-all attributes is unnecessary. Or maybe the application uses several layers
-that share the same attributes (typical when using multiple layers to
-create effects like shadows or highlights around the primary primitives),
-making it desirable to avoid having multiple layers repeat the same
-automatic attribute recalculations.
-
-This allows an application to supply its own buffers properties.
-If attributes are shared between layers, a useful technique is to create
-a separate AttributeManager and calculate the attributes once, and then
-pass them in as "overrides" to the layers.
+This prop expects an object with key/value pairs that the keys matching attribute names.
+If any values in this object are changed between props udpate, the corresponding attribute
+will be invalidated.
+
+Using this method, the attribute update can happen when the content of `data` prop changed, or
```

Comment:
Maybe drop the "on the GPU side"? Simpler is better?

Change: can happen not only when ..., but also ...

Final statement: updateTriggers cannot block attribute updates, just trigger them. Overriding updateState can block attribute updates.